symengine/ntheory.cpp

```diff
@@ -1474,4 +1474,74 @@ void powermod_list(std::vector<RCP<const Integer>> &pows, const RCP<const Intege
     }
 }
 
+std::vector<mpz_class> quadratic_residues(const Integer &a)
+{
+    /*
+        Returns the list of quadratic residues.
+        Example
+        ========
+        >>> quadratic_residues(7)
+        [0, 1, 2, 4]
+    */
+
+    if (a.as_mpz() < 1) {
+        throw std::runtime_error("quadratic_residues: Input must be > 0");
+    }
+
+    std::vector<mpz_class> residue;
+    for( mpz_class i = 0; i <= a.as_int()/2; i++)
+    {
+        residue.push_back( (i*i) % a.as_int());
+    }
+
+    sort( residue.begin(), residue.end() );
+    residue.erase( unique( residue.begin(), residue.end() ), residue.end() );
+
+    return residue;
+
+}
+
+bool is_quad_residue(const Integer &a , const Integer &p)
+{
+    /*
+    Returns true if ``a`` (mod ``p``) is in the set of squares mod ``p``,
+    i.e a % p in set([i**2 % p for i in range(p)]). If ``p`` is an odd but
+    not prime, an iterative method is used to make the determination.
+    */
+
+
+    if(p.as_mpz() < 1)
+        throw std::runtime_error("is_quad_residue: Second parameter must be > 0");
+
+    mpz_class a_final;
+    a_final = a.as_mpz();
+    if(a.as_mpz() >= p.as_mpz() || a.as_mpz() < 0)
+        a_final = a.as_mpz() % p.as_mpz();
+    if(a_final < 0 && a_final > -p.as_mpz())
+        a_final += p.as_mpz();
+    if(a_final < 2 || p.as_mpz() < 3)
+        return true;
+
+    if(!probab_prime_p(*integer(p.as_mpz()),0))
+     {
+        if((p.as_mpz() % 2 == 1 ) && jacobi(*integer(a_final),p) == -1)
+            return false;
+
+        RCP<const Integer> x;
+        const RCP<const Integer> a1 = integer(a_final);
+        const RCP<const Integer> p1 = integer(p.as_mpz());
+
+        bool r = true;
+        r = nthroot_mod(outArg(x), a1, integer(2), p1);
+        return r;
+    }
+
+    mpz_t expo;
+    mpz_init(expo);
+    mpz_pow_ui(expo, a_final.get_mpz_t(), (p.as_mpz().get_ui()-1)/2);
```

Comment:
In your current code, I think you actually have a memory leak, since you never call `mpz_clear(expo)`... 
Do not use `mpz_t`. Use `mpz_class` instead, that fixes all memory issues.

So do this as:
```
mpz_class expo;
mpz_pow_ui(expo.get_mpz_t(), a_final.get_mpz_t(), (p.as_mpz().get_ui()-1)/2);
```